### ISSUE TITLE
code 400

### ISSUE THREAD

**Issue by ldc2726** [semantic-release/gitlab#239]
021-05-20T05:02:43.750Z semantic-release:git commited files: [ 'CHANGELOG.md' ]
[1:02:44 PM] [semantic-release] [@semantic-release/git] › ℹ  Prepared Git release: v1.2.5
[1:02:44 PM] [semantic-release] › ✔  Completed step "prepare" of plugin "@semantic-release/git"
[1:02:44 PM] [semantic-release] › ℹ  Start step "generateNotes" of plugin "@semantic-release/release-notes-generator"
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator version: '1.2.5'
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator host: undefined
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator owner: 'dyWeb'
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator repository: 'DataCockpit/demo'
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator previousTag: 'v1.2.4'
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator currentTag: 'v1.2.5'
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator host: 'http://192.168.1.184'
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator linkReferences: undefined
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator issue: 'issues'
2021-05-20T05:02:44.648Z semantic-release:release-notes-generator commit: 'commit'
[1:02:44 PM] [semantic-release] › ✔  Completed step "generateNotes" of plugin "@semantic-release/release-notes-generator"
[1:02:46 PM] [semantic-release] › ✔  Created tag v1.2.5
[1:02:46 PM] [semantic-release] › ℹ  Start step "publish" of plugin "@semantic-release/gitlab"
2021-05-20T05:02:46.971Z semantic-release:gitlab repoId: 'dyWeb/DataCockpit/demo'
2021-05-20T05:02:46.971Z semantic-release:gitlab release name: 'v1.2.5'
2021-05-20T05:02:46.971Z semantic-release:gitlab release ref: '852ee054f6c68bab2084651f7598d2f74ad2face'
2021-05-20T05:02:46.971Z semantic-release:gitlab Create a release for git tag 'v1.2.5' with commit '852ee054f6c68bab2084651f7598d2f74ad2face'
gitlabApiUrl: http://192.168.1.184/api/v4 /projects/dyWeb%2FDataCockpit%2Fdemo/release
apiOptions: { headers: { 'PRIVATE-TOKEN': '[secure]' } }
v1.2.5 ## [1.2.5](http://192.168.1.184/dyWeb/DataCockpit/demo/compare/v1.2.4...v1.2.5) (2021-05-20)

### Bug Fixes

* 分开一个分之测试 ([30ce41a](http://192.168.1.184/dyWeb/DataCockpit/demo/commit/30ce41a0174dc1c708df45ff141a7e57a68926ea))

 []
[1:02:47 PM] [semantic-release] › ✖  Failed step "publish" of plugin "@semantic-release/gitlab"
[1:02:47 PM] [semantic-release] › ✖  An error occurred while running semantic-release: HTTPError: Response code 400 (Bad Request)
    at EventEmitter.<anonymous> (/usr/local/lib/node_modules/@semantic-release/gitlab/node_modules/got/dist/source/as-promise.js:118:31)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  pluginName: '@semantic-release/gitlab'
}
HTTPError: Response code 400 (Bad Request)
    at EventEmitter.<anonymous> (/usr/local/lib/node_modules/@semantic-release/gitlab/node_modules/got/dist/source/as-promise.js:118:31)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  pluginName: '@semantic-release/gitlab'
}ERROR: Job failed: exit status 1

**Comment by ldc2726** [comment]
{
    "branches": [
        "+([0-9])?(.{+([0-9]),x}).x", 
        "master", 
        "next",
        {"name": "beta", "prerelease": true}, 
        {"name": "alpha", "prerelease": true}
    ],
    "plugins": [
        "@semantic-release/commit-analyzer",
        "@semantic-release/release-notes-generator",
        "@semantic-release/changelog",
        [
            "@semantic-release/gitlab",
            {
                "assets": [
                    { "path": "dist/*.html", "label": "html distribution" }
                ]
            }
        ],
        [
            "@semantic-release/git",
            {
                "assets": [
                    "package.json",
                    "CHANGELOG.md"
                ],
                "message": "chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
            }
        ],
    ]
}
![image](https://user-images.githubusercontent.com/34648806/118975152-451a6700-b9a6-11eb-901a-2189c1dbd5c8.png)

**Comment by ldc2726** [comment]
![image](https://user-images.githubusercontent.com/34648806/118975222-5cf1eb00-b9a6-11eb-8462-c7e342cf4b49.png)
{
    "branches": [
        "+([0-9])?(.{+([0-9]),x}).x", 
        "master", 
        "next",
        {"name": "beta", "prerelease": true}, 
        {"name": "alpha", "prerelease": true}
    ],
    "plugins": [
        "@semantic-release/commit-analyzer",
        "@semantic-release/release-notes-generator",
        "@semantic-release/changelog",
        [
            "@semantic-release/gitlab",
            {
                "assets": [
                    { "path": "dist/*.html", "label": "html distribution" }
                ]
            }
        ],
        [
            "@semantic-release/git",
            {
                "assets": [
                    "package.json",
                    "CHANGELOG.md"
                ],
                "message": "chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
            }
        ],
    ]
}

**Comment by ldc2726** [comment]
help,,,,,,why????

**Comment by ldc2726** [comment]
![image](https://user-images.githubusercontent.com/34648806/118975331-7dba4080-b9a6-11eb-8747-a427e9354bf8.png)
`{
    "branches": [
        "+([0-9])?(.{+([0-9]),x}).x", 
        "master", 
        "next",
        {"name": "beta", "prerelease": true}, 
        {"name": "alpha", "prerelease": true}
    ],
    "plugins": [
        "@semantic-release/commit-analyzer",
        "@semantic-release/release-notes-generator",
        "@semantic-release/changelog",
        [
            "@semantic-release/gitlab",
            {
                "assets": [
                    { "path": "dist/*.html", "label": "html distribution" }
                ]
            }
        ],
        [
            "@semantic-release/git",
            {
                "assets": [
                    "package.json",
                    "CHANGELOG.md"
                ],
                "message": "chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
            }
        ],
    ]
}
`

**Comment by ldc2726** [comment]
![image](https://user-images.githubusercontent.com/34648806/118975411-91fe3d80-b9a6-11eb-87df-baf9e6576bda.png)
{
    "branches": [
        "+([0-9])?(.{+([0-9]),x}).x", 
        "master", 
        "next",
        {"name": "beta", "prerelease": true}, 
        {"name": "alpha", "prerelease": true}
    ],
    "plugins": [
        "@semantic-release/commit-analyzer",
        "@semantic-release/release-notes-generator",
        "@semantic-release/changelog",
        [
            "@semantic-release/gitlab",
            {
                "assets": [
                    { "path": "dist/*.html", "label": "html distribution" }
                ]
            }
        ],
        [
            "@semantic-release/git",
            {
                "assets": [
                    "package.json",
                    "CHANGELOG.md"
                ],
                "message": "chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
            }
        ],
    ]
}

**Comment by ldc2726** [comment]
before:
![image](https://user-images.githubusercontent.com/34648806/119074035-de865f00-ba20-11eb-92b1-3f508a185f56.png)
add "name":
![image](https://user-images.githubusercontent.com/34648806/119074004-d3333380-ba20-11eb-8d8b-2dbb5eaac597.png)
after：
![image](https://user-images.githubusercontent.com/34648806/119073956-c0206380-ba20-11eb-9d0c-814cf611a815.png)

I tried to add the attribute "name" to the body, and the result was successful. I hope to optimize it as soon as possible, I think it may be a bug, thank you

**Comment by nfriend** [comment]
@ldc2726 Thanks for reporting! Very weird. [According to the docs](https://docs.gitlab.com/ee/api/releases/#create-a-release), the `name` attribute is optional - if it's omitted, it defaults to the `tag_name` property.

I just confirmed this myself by `POST`ing to a test project on GitLab.com with the following body:

```json
{
    "tag_name": "v5.5"
}
```

and the release was successfully created: https://gitlab.com/nfriend/release-test/-/releases/v5.5

But your screenshots make it very clear that the missing `name` property is causing this error 🤔 

What version of GitLab are using? Also, what permissions do you have on the project? (Developer, Maintainer, Owner, etc.)

**Comment by ldc2726** [comment]
@nfriend 
1.Owner
2.
![image](https://user-images.githubusercontent.com/34648806/119296803-3d9bdc00-bc8c-11eb-95c9-075e2f35ff53.png)

No matter what version is causing the problem, I still want to be as compatible as possible by adding the "name" attribute to it，thank you!

**Comment by ldc2726** [comment]
@nfriend 
1.Owner
2.
![image](https://user-images.githubusercontent.com/34648806/119296803-3d9bdc00-bc8c-11eb-95c9-075e2f35ff53.png)

No matter what version is causing the problem, I still want to be as compatible as possible by adding the "name" attribute to it，thank you!

**Comment by nfriend** [comment]
@ldc2726 Ah, yep, the GitLab version is the culprit!

I looks like this attribute [was required in GitLab 11.X](https://docs.gitlab.com/11.11/ee/api/releases/index.html#create-a-release), but [then changed to be optional in GitLab 12.X](https://docs.gitlab.com/12.10/ee/api/releases/index.html#create-a-release).

I'm a little unsure what to do here, since according to [Gitlab's support schedule](https://about.gitlab.com/support/statement-of-support.html#we-support-the-current-major-version-and-the-two-previous-major-versions), 11.X will become unsupported in 28 days (when GitLab 14.0 is released).

The _best_ solution is to upgrade from 11.X. But I also understand this is a big undertaking and maybe not within your control.

If upgrading is really not a solution for you, you could:

1. Submit a PR that adds the `name` attribute to the request
1. Include a comment above the line that indicates this is only for compatibility with an unsupported version of GitLab
1. Update the docs with an explicit list of all supported GitLab versions

WDYT?

**Comment by ldc2726** [comment]
Well, I chose to submit a PR so that more people could avoid this problem.
#240

**Comment by nfriend** [comment]
Thanks @ldc2726! I won't have time to review this week but will prioritize this next week!

**Comment by alecgerona** [comment]
Also seems to happen on Gitlab.com.
![image](https://user-images.githubusercontent.com/6406447/121480371-947c1200-c9fd-11eb-90db-b5d41864b0f3.png)

Same 400 error code but not really sure if the same cause of failure since the reason is not explicitly logged.

**Comment by nfriend** [comment]
@joffreyvillard That's very possible! Are you using the latest version of `@semantic-release/gitlab`? This plugin was updated to stop using the deprecated API in https://github.com/semantic-release/gitlab/pull/184, which is included in version 6.0.7 and beyond.

**Comment by nfriend** [comment]
> Same 400 error code but not really sure if the same cause of failure since the reason is not explicitly logged.

@alecgerona I agree explicitly logging the error code is a good step 👍 I'll try and open a PR for this tomorrow.

And sorry for the slow turnaround time here, @ldc2726 - I'm hoping to revisit this tomorrow!

**Comment by joffreyvillard** [comment]
Thanks for your reply @nfriend!

I'm using `semantic-release/gitlab` via the `aplyca/semantic-release` Docker image (https://hub.docker.com/r/aplyca/semantic-release) which is 8 months old, and embeds `semantic-release` version 17.4.4 (weirdly enough, the logs say 17.1.2) and `@semantic-release/gitlab` version 6.1.0

![image](https://user-images.githubusercontent.com/37925755/122359456-dce38480-cf55-11eb-8663-aa70631f37a8.png)

I guess I could start by building that image myself with the latest versions from npm.

**Comment by nfriend** [comment]
Just FYI, I opened a PR to add better logging: https://github.com/semantic-release/gitlab/pull/245. Once this is merged, hopefully it will be a little easier to debug the root issue from the logs.

**Comment by joffreyvillard** [comment]
I can confirm that it works well on my side with the latest version (using `node:alpine` image + `npm install -g semantic-release @semantic-release/gitlab ...` instead of `aplyca/semantic-release`) - don't get why the `npm view` commands run within a container using that last image were indicating that the last version was in use :shrug: 

Thanks again for your prompt replies and the suggested/ongoing improvements

**Comment by nfriend** [comment]
@joffreyvillard Awesome, glad to hear it!

**Comment by EvanCarroll** [comment]
@joffreyvillard  that's because `npm view` shows you the version on the repo, not the version you have installed. You want

```sh
/ # npm list -g semantic-release @semantic-release/gitlab 
/usr/local/lib
+-- @semantic-release/gitlab@6.2.1
| `-- semantic-release@17.4.4 deduped
`-- semantic-release@17.4.4
  +-- @semantic-release/commit-analyzer@8.0.1
  | `-- semantic-release@17.4.4 deduped
  +-- @semantic-release/github@7.2.3
  | `-- semantic-release@17.4.4 deduped
  +-- @semantic-release/npm@7.1.3
  | `-- semantic-release@17.4.4 deduped
  `-- @semantic-release/release-notes-generator@9.0.3
    `-- semantic-release@17.4.4 deduped
```

**Comment by EvanCarroll** [comment]
I also don't think @joffreyvillard and @ldc2726's issues are the same. @joffreyvillard pasted an error with a 404. This issue isolates error 400.

**Comment by nfriend** [comment]
Thanks everyone for the discussion!

It's been determined the root cause of this specific 400 error is an outdated GitLab version (v11) - [see this comment above](https://github.com/semantic-release/gitlab/issues/239#issuecomment-847924888). We've decided not to fix this issue to minimize the maintenance burden of this plugin and stay consistent with this plugin's [compatibility guidelines](https://github.com/semantic-release/gitlab/#compatibility). [More information about this decision here](https://github.com/semantic-release/gitlab/pull/240#issuecomment-893672094).

I'm going to go ahead and close this issue. However, please feel free to open a new issue if you are still seeing 400 errors on GitLab version 12 or later.

As a side note, errors like this should be easier to debug in the latest version of `@semantic-release/gitlab` since it now prints out more debug information (see https://github.com/semantic-release/gitlab/pull/245).
